File: wordgen/generate.go
package wordgen

import (
	"errors"
	"math"
	"math/rand"
	"strings"
)

// Draw randomly draws a string from a distribution.
func (distr *Distribution) Draw(random *rand.Rand) (string, float64) {
	part := random.Intn(distr.weightSum)
	for i, weight := range distr.weights {
		part -= weight
		if part <= 0 {
			likelihood := float64(weight) / float64(distr.weightSum)
			return distr.candidates[i], likelihood
		}
	}
	panic("invalid distribution")
}

// SetPrefix fixes the prefix of a generted word.
func (model *Model) SetPrefix(prefix string) error {
	// Basically, we use the ending n-gram of specified prefix as the initial
	// state of a Markov chain. Here, we need to deal with these edge cases:
	//
	// - The prefix is shorter than n characters.
	// - The ending n-gram did not appear in training data (unknown n-gram).
	//
	// We deal with the issues by capping the end of the prefix with known
	// n-grams. The result is a distribution of capped prefixes.

	prefix = model.metadata.Prefix + prefix

	cutpos := len(prefix) - model.metadata.Ngram
	if cutpos < 0 {
		cutpos = 0
	}

	for ; cutpos < len(prefix); cutpos++ {
		prefixStem := prefix[:cutpos]
		prefixEnd := prefix[cutpos:]

		candidates := []string{}
		counts := []int{}

		for ngram, distr := range model.transitions {
			if strings.HasPrefix(ngram, prefixEnd) {
				candidates = append(candidates, prefixStem+ngram)
				counts = append(counts, distr.weightSum)
			}
		}

		if len(candidates) > 0 {
			model.prefix, _ = makeDistribution(Histogram{
				Candidates: candidates,
				Counts:     counts,
			})
			return nil
		}
	}

	return errors.New("prefix not recognizable")
}

// Generate randomly generates a word-like string. Returns a generated word and
// the likelihood of the word.
func (model *Model) Generate(random *rand.Rand) (string, float64) {
	word, lik := model.prefix.Draw(random)
	logLik := math.Log(lik)
	steps := 0

	for !strings.HasSuffix(word, model.metadata.Suffix) {
		tail := len(word) - model.metadata.Ngram
		distr := model.transitions[word[tail:]]
		nextChar, lik := distr.Draw(random)

		word += nextChar
		logLik += math.Log(lik)

		steps++
	}

	word = strings.TrimPrefix(word, model.metadata.Prefix)
	word = strings.TrimSuffix(word, model.metadata.Suffix)

	meanLik := math.Exp(logLik / float64(steps))

	return word, meanLik
}

File: wordgen/generate.py
import random

from .model import PREFIX


def generate(model, prefix=None, max_length=64):
    if prefix is None:
        return generate_normal(model, max_length)
    else:
        return generate_with_prefix(model, prefix, max_length)


def generate_normal(model, max_length):
    tokens = model.tokens
    initial_states, initial_weights = model.prior

    while True:
        state, = random.choices(initial_states, initial_weights)
        word = tokens[state]

        while len(word) < max_length + 2:
            destinations, weights = model.posterior(state)
            if len(destinations) == 0:
                break
            state, = random.choices(destinations, weights)
            word += tokens[state][-1]

        # Trim PREFIX and SUFFIX.
        word = word[1:-1]

        yield word


def generate_with_prefix(model, prefix, max_length):
    tokens = model.tokens
    prefix = PREFIX + prefix

    prefix_state = get_prefix_state(model, prefix)
    if prefix_state is None:
        return

    while True:
        state = prefix_state
        word = prefix

        while len(word) < max_length + 2:
            destinations, weights = model.posterior(state)
            if len(destinations) == 0:
                break
            state, = random.choices(destinations, weights)
            word += tokens[state][-1]

        # Trim PREFIX and SUFFIX.
        word = word[1:-1]

        yield word


def get_prefix_state(model, prefix):
    tokens = model.tokens

    if len(prefix) >= model.token_size:
        prefix_token = prefix[-model.token_size:]
        try:
            return tokens.index(prefix_token)
        except ValueError:
            return None

    initial_states, initial_weights = model.prior
    prefix_states = []
    prefix_weights = []
    for state, weight in zip(initial_states, initial_weights):
        if tokens[state].startswith(prefix):
            prefix_states.append(state)
            prefix_weights.append(weight)

    if len(prefix_states) == 0:
        return None
    prefix_state, = random.choices(prefix_states, prefix_weights)

    return prefix_state
